packages/eslint-plugin/src/rules/prefer-return-this-type.ts

```diff
@@ -0,0 +1,161 @@
+import {
+  TSESTree,
+  AST_NODE_TYPES,
+} from '@typescript-eslint/experimental-utils';
+import { createRule, forEachReturnStatement, getParserServices } from '../util';
+import * as ts from 'typescript';
+
+const IgnoreTypes = new Set<TSESTree.TypeNode['type']>([
+  AST_NODE_TYPES.TSThisType,
+  AST_NODE_TYPES.TSAnyKeyword,
+  AST_NODE_TYPES.TSUnknownKeyword,
+]);
+
+type ClassLikeDeclaration =
+  | TSESTree.ClassDeclaration
+  | TSESTree.ClassExpression;
+
+type FunctionLike =
+  | TSESTree.MethodDefinition['value']
+  | TSESTree.ArrowFunctionExpression;
+
+export default createRule({
+  name: 'prefer-return-this-type',
+  defaultOptions: [],
+
+  meta: {
+    type: 'suggestion',
+    docs: {
+      description:
+        'Enforce that `this` is used when only `this` type is returned',
+      category: 'Best Practices',
+      recommended: 'error',
+      requiresTypeChecking: true,
+    },
+    messages: {
+      UseThisType: 'use `this` type instead.',
```

Comment:
nit - convention is to use camelCase
```suggestion
      useThisType: 'use `this` type instead.',
```